src/WebJobs.Script/Host/FunctionMetadataManager.cs

```diff
@@ -18,36 +21,81 @@ namespace Microsoft.Azure.WebJobs.Script
-        private readonly ILogger _logger;
+        private bool _isHttpWorker;
+        private bool _servicesReset = false;
+        private ILogger _logger;
+        private IOptions<ScriptJobHostOptions> _scriptOptions;
+        private ImmutableArray<FunctionMetadata> _functionMetadataArray;
+        private IEnumerable<IFunctionProvider> _functionProviders;
         private Dictionary<string, ICollection<string>> _functionErrors = new Dictionary<string, ICollection<string>>();
 
-        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IOptions<HttpWorkerOptions> httpWorkerOptions, ILoggerFactory loggerFactory)
+        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IEnumerable<IFunctionProvider> functionProviders, IOptions<HttpWorkerOptions> httpWorkerOptions, IScriptHostManager scriptHostManager, ILoggerFactory loggerFactory)
         {
             _scriptOptions = scriptOptions;
+            _serviceProvider = scriptHostManager as IServiceProvider;
             _functionMetadataProvider = functionMetadataProvider;
             _logger = loggerFactory.CreateLogger(LogCategories.Startup);
-            _functionMetadataArray = new Lazy<ImmutableArray<FunctionMetadata>>(LoadFunctionMetadata);
-            _isHttpWorker = httpWorkerOptions.Value.Description != null;
-        }
+            _isHttpWorker = httpWorkerOptions?.Value?.Description != null;
+            _functionProviders = functionProviders;
 
-        public ImmutableArray<FunctionMetadata> Functions => _functionMetadataArray.Value;
+            scriptHostManager.HostInitializing += ResetScriptHostServices;
+        }
 
         public ImmutableDictionary<string, ImmutableArray<string>> Errors { get; private set; }
 
+        public ImmutableArray<FunctionMetadata> GetFunctionMetadata(bool forceRefresh)
+        {
+            if (!forceRefresh && !_servicesReset && !_functionMetadataArray.IsDefaultOrEmpty)
+            {
+                return _functionMetadataArray;
+            }
+
+            _functionMetadataArray = LoadFunctionMetadata(forceRefresh);
+            _servicesReset = false;
+            return _functionMetadataArray;
+        }
+
+        private void ResetScriptHostServices(object sender, EventArgs e)
+        {
+            _functionProviders = _serviceProvider?.GetService<IEnumerable<IFunctionProvider>>();
+            _isHttpWorker = _serviceProvider?.GetService<IOptions<HttpWorkerOptions>>()?.Value?.Description != null;
+            _scriptOptions = _serviceProvider?.GetService<IOptions<ScriptJobHostOptions>>();
+
+            // Resetting the logger switches the logger scope to Script Host level,
+            // also making the logs available to Application Insights
+            _logger = _serviceProvider?.GetService<ILoggerFactory>().CreateLogger(LogCategories.Startup);
```

Comment:
I want to explicitly call this out, to be sure it makes sense and is ok.

Reply:
Not sure I understand this "reset services" concept

Reply:
FunctionMetadataManager is created now at the scope of the WebHost. This means that when it is initialized, the `IFunctionProvider`s would not be available yet. These providers come from the extensions controlled by the user code. So even though `FunctionMetadataManager` is at WebHost level, we need to "reset" these services everytime JobHost starts. This way we make sure any code changes reflect the proper metadata.